File: client/Assets/Scripts/Sdk/Sdk.cs
using System;
using System.Collections.Generic;
using System.Linq;
using NovelCraft.Utilities.Logger;
using NovelCraft.Utilities.Messages;

namespace NovelCraft.Sdk
{

    /// <summary>
    /// The SDK.
    /// </summary>
    public static partial class Sdk
    {
        private const int GetInfoInterval = 1000;
        private const int PingInterval = 1000;


        /// <summary>
        /// Gets the agent representing the player controlled by the user.
        /// </summary>
        public static IAgent? Agent => _agent;

        /// <summary>
        /// Gets the block collection.
        /// </summary>
        public static IBlockSource? Blocks => _blockSource;

        /// <summary>
        /// Gets the client for sending and receiving messages directly to or from the server.
        /// </summary>
        public static IClient? Client => _client;

        /// <summary>
        /// Gets the list of all entities in the world.
        /// </summary>
        public static IEntitySource? Entities => _entitySource;

        /// <summary>
        /// Gets the logger.
        /// </summary>
        public static ILogger Logger => _userLogger;

        /// <summary>
        /// Gets the latency between the client and the server.
        /// </summary>
        public static TimeSpan? Latency => _latency;

        /// <summary>
        /// Gets current tick.
        /// </summary>
        public static int? Tick
        {
            get
            {
                if (_lastTickInfo is null)
                {
                    return null;
                }

                if (TicksPerSecond is null)
                {
                    return _lastTickInfo.Value.LastTick;
                }

                var timeSinceLastTick = DateTime.Now - _lastTickInfo.Value.LastTickTime;
                var ticksSinceLastTick = (int)((decimal)timeSinceLastTick.TotalSeconds * TicksPerSecond);

                return _lastTickInfo.Value.LastTick + ticksSinceLastTick;
            }
        }

        /// <summary>
        /// Gets the number of ticks per second.
        /// </summary>
        public static decimal? TicksPerSecond { get; private set; } = null;


        internal static Agent? _agent = null;
        internal static BlockSource? _blockSource = null;
        internal static Client? _client = null;
        internal static EntitySource? _entitySource = null;
        internal static TimeSpan? _latency = null;
        internal static ILogger _sdkLogger { get; } = new Logger("SDK");
        internal static ILogger _userLogger { get; } = new Logger("User");

        private static (int LastTick, DateTime LastTickTime)? _lastTickInfo = null;
        private static System.Timers.Timer _getInfoTimer = new(GetInfoInterval);
        private static System.Timers.Timer _pingTimer = new(PingInterval);
        private static string? _token = null;

        public static void CloseTimers()
        {
            _pingTimer.Close();
            _getInfoTimer.Close();
        }
        /// <summary>
        /// Initializes the SDK.
        /// </summary>
        /// <param name="config">The configuration of the SDK.</param>
        public static void Initialize(string token, string host, int port)
        {
            try
            {
                _sdkLogger.Info("Initializing SDK...");

                _token = token;

                // Initialize the client
                _client = new Client(host, port);
                _client.AfterMessageReceiveEvent += OnAfterMessageReceiveEvent;

                // Manually call the tick event to get information right away.
                OnTick();

                // Initialize the timers.
                _getInfoTimer.Elapsed += (sender, e) => OnTick();
                _getInfoTimer.Start();

                _pingTimer.Elapsed += (sender, e) =>
                {
                    Client?.Send(new ClientPingMessage()
                    {
                        Token = _token!,
                        SentTime = (decimal)DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond
                    });
                };
                _pingTimer.Start();

            }
            catch (Exception e)
            {
                throw new Exception($"Failed to initialize SDK: {e.Message}", e);
            }
        }

        private static void OnAfterMessageReceiveEvent(object? sender, IMessage message)
        {
            switch (message)
            {
                case ServerPongMessage msg:
                    _latency = TimeSpan.FromMilliseconds((double)(DateTime.Now.Ticks - (long)(msg.SentTime * TimeSpan.TicksPerMillisecond)) / TimeSpan.TicksPerMillisecond);
                    break;


                case ErrorMessage msg:
                    _sdkLogger.Error($"The server returned an error: {msg.Message} ({msg.Code})");
                    break;


                case ServerGetBlocksAndEntitiesMessage msg:
                    // Update the block source.
                    if (_blockSource is null)
                    {
                        _blockSource = new BlockSource();
                    }

                    _blockSource.Clear();
                    foreach (var section in msg.Sections)
                    {
                        _blockSource.AddSection(
                          new Section(
                            new Position<int>(section.Position.X, section.Position.Y, section.Position.Z),
                            section.Blocks
                          )
                        );
                    }

                    // Update the entity source.
                    if (_entitySource is null)
                    {
                        _entitySource = new EntitySource();
                    }

                    _entitySource.Clear();
                    foreach (var entity in msg.Entities)
                    {
                        _entitySource[entity.UniqueId] = new Entity(entity.UniqueId, entity.TypeId,
                          new Position<decimal>(entity.Position.X, entity.Position.Y, entity.Position.Z),
                          new Orientation(entity.Orientation.Yaw, entity.Orientation.Pitch));

                        if (_agent is not null && entity.UniqueId == _agent.UniqueId)
                        {
                            _agent.Position = new Position<decimal>(entity.Position.X, entity.Position.Y, entity.Position.Z);
                            _agent.Orientation = new Orientation(entity.Orientation.Yaw, entity.Orientation.Pitch);
                        }
                    }
                    break;


                case ServerGetPlayerInfoMessage msg:
                    if (_token is null)
                    {
                        throw new InvalidOperationException("The token is null.");
                    }

                    if (_agent is null)
                    {
                        _agent = new Agent(_token, msg.UniqueId,
                          new Position<decimal>()
                          {
                              X = msg.Position.X,
                              Y = msg.Position.Y,
                              Z = msg.Position.Z
                          },
                          new Orientation()
                          {
                              Yaw = msg.Orientation.Yaw,
                              Pitch = msg.Orientation.Pitch
                          });
                    }

                    if (msg.Inventory.Count != _agent.Inventory.Size)
                    {
                        throw new InvalidOperationException("The inventory size of the agent is not equal to the inventory size of the server.");
                    }

                    if (msg.UniqueId != _agent.UniqueId)
                    {
                        throw new InvalidOperationException("The unique ID of the agent is not equal to the unique ID of the server.");
                    }

                    _agent.Health = msg.Health;
                    _agent.Orientation = new Orientation(msg.Orientation.Yaw, msg.Orientation.Pitch);
                    _agent.Position = new Position<decimal>(msg.Position.X, msg.Position.Y, msg.Position.Z);
                    _agent._inventory.MainHandSlot = msg.MainHand;
                    for (int i = 0; i < msg.Inventory.Count; i++)
                    {
                        var itemInfo = msg.Inventory[i];
                        _agent._inventory[i] = (itemInfo is null) ? null : new ItemStack(itemInfo.TypeId, itemInfo.Count);
                    }
                    break;


                case ServerGetTickMessage serverGetTickMessage:
                    _lastTickInfo = (serverGetTickMessage.Tick, DateTime.Now);
                    TicksPerSecond = serverGetTickMessage.TicksPerSecond;
                    break;


                case ServerAfterBlockChangeMessage msg:
                    // To be tested
                    if (Blocks is not null)
                    {
                        foreach (var blockChangeInfo in msg.ChangeList)
                        {
                            Position<int> blockPosition = new Position<int>(blockChangeInfo.Position.X, blockChangeInfo.Position.Y, blockChangeInfo.Position.Z);
                            ((BlockSource)Blocks)[blockPosition] = new Block(blockChangeInfo.BlockTypeId, blockPosition);
                        }
                    }
                    break;


                case ServerAfterEntityCreateMessage msg:
                    if (Entities is not null)
                    {
                        foreach (var entityCreateInfo in msg.CreationList)
                        {
                            EntitySource entitySource = (EntitySource)Entities;

                            // Have checked if the key exists in the function "AddEntity"
                            entitySource.AddEntity(new Entity(entityCreateInfo.UniqueId, entityCreateInfo.EntityTypeId,
                            new Position<decimal>(entityCreateInfo.Position.X, entityCreateInfo.Position.Y, entityCreateInfo.Position.Z),
                            new Orientation(entityCreateInfo.Orientation.Yaw, entityCreateInfo.Orientation.Pitch)));
                        }
                    }
                    break;


                case ServerAfterEntityRemoveMessage msg:
                    if (Entities is not null)
                    {
                        foreach (int removalId in msg.RemovalIdList)
                        {
                            // Have checked if the key exists in the function "RemoveEntity"
                            ((EntitySource)Entities).RemoveEntity(removalId);
                        }
                    }
                    break;


                case ServerAfterEntityOrientationChangeMessage msg:
                    if (Entities is not null)
                    {
                        foreach (var changeInfo in msg.ChangeList)
                        {
                            var entity = Entities[changeInfo.UniqueId];
                            if (entity is not null)
                            {
                                ((EntitySource)Entities)[changeInfo.UniqueId] = new Entity(entity.UniqueId, entity.TypeId, entity.Position,
                                new Orientation(changeInfo.Orientation.Yaw, changeInfo.Orientation.Pitch));
                            }

                            if (_agent is not null && changeInfo.UniqueId == _agent.UniqueId)
                            {
                                _agent.Orientation = new Orientation(changeInfo.Orientation.Yaw, changeInfo.Orientation.Pitch);
                            }
                        }
                    }
                    break;


                case ServerAfterEntityPositionChangeMessage msg:
                    if (Entities is not null)
                    {
                        foreach (var changeInfo in msg.ChangeList)
                        {
                            var entity = Entities[changeInfo.UniqueId];
                            if (entity is not null)
                            {
                                ((EntitySource)Entities)[changeInfo.UniqueId] = new Entity(entity.UniqueId, entity.TypeId,
                                new Position<decimal>(changeInfo.Position.X, changeInfo.Position.Y, changeInfo.Position.Z),
                                entity.Orientation, new Position<decimal>(changeInfo.Velocity.X, changeInfo.Velocity.Y, changeInfo.Velocity.Z));
                            }

                            if (_agent is not null && changeInfo.UniqueId == _agent.UniqueId)
                            {
                                _agent.Position = new Position<decimal>(changeInfo.Position.X, changeInfo.Position.Y, changeInfo.Position.Z);
                                _agent.Velocity = new Position<decimal>(changeInfo.Velocity.X, changeInfo.Velocity.Y, changeInfo.Velocity.Z);
                            }
                        }
                    }
                    break;


                case ServerAfterPlayerInventoryChangeMessage msg:
                    // TO DO: Implement this case
                    break;
            }
        }

        private static void OnTick()
        {
            string token = _token ?? throw new InvalidOperationException("The SDK is not initialized.");

            Client?.Send(new ClientGetTickMessage()
            {
                Token = token
            });

            if (Agent is not null)
            {
                List<NovelCraft.Utilities.Messages.Position<int>> requestPositionList = (
                  from x in Enumerable.Range(-1, 3)
                  from y in Enumerable.Range(-1, 3)
                  from z in Enumerable.Range(-1, 3)
                  select new NovelCraft.Utilities.Messages.Position<int>()
                  {
                      X = (int)(Agent.Position.X + x * 16),
                      Y = (int)(Agent.Position.Y + y * 16),
                      Z = (int)(Agent.Position.Z + z * 16)
                  }
                ).ToList();

                Client?.Send(new ClientGetBlocksAndEntitiesMessage()
                {
                    Token = token,
                    RequestSectionList = requestPositionList
                });
            }

            Client?.Send(new ClientGetPlayerInfoMessage()
            {
                Token = token
            });
        }
    }

}

File: sdks/python/src/novelcraft/sdk/sdk.py
"""The SDK APIs.

Here is an example of how to use the SDK APIs:
```python
import novelcraft.sdk as sdk

agent = sdk.get_agent()
```
"""

from __future__ import annotations

import argparse
import asyncio
from dataclasses import dataclass
from datetime import datetime

from .agent import IAgent, Agent
from .block import Block
from .block_source import IBlockSource, BlockSource
from .client import Client
from .entity import Entity
from .entity_source import IEntitySource, EntitySource
from .item_stack import ItemStack
from .logger import ILogger, Logger
from .message import Message
from .orientation import Orientation
from .position import Position
from .section import Section
from .inventory import Inventory


@dataclass
class _TickInfo:
    """The tick info."""
    tick: int
    time: datetime


GET_INFO_INTERVAL = 10
PING_INVERVAL = 10

_agent: Agent | None = None
_block_source: BlockSource | None = None
_client: Client | None = None
_entity_source: EntitySource | None = None
_tick_info: _TickInfo | None = None
_task_list: list[asyncio.Task] = []
_token: str | None = None
_tps: float | None = None
_sdk_logger: Logger = Logger('SDK')
_user_logger: Logger = Logger('User')


async def finalize() -> None:
    """Finalizes the SDK.

    This function should be called when the program is about to exit.
    """
    _sdk_logger.info('Finalizing SDK...')

    if _client is not None:
        await _client.stop()

    for task in _task_list:
        task.cancel()


async def initialize() -> None:
    """Initializes the SDK.

    This function should be called before any other SDK APIs are used.
    """
    try:
        global _token, _client

        _sdk_logger.info('Initializing SDK...')

        parser = argparse.ArgumentParser()
        parser.add_argument('--host', type=str, default='localhost')
        parser.add_argument('--port', type=int, default=14514)
        parser.add_argument('--token', type=str, default='')
        args = parser.parse_args()

        _client = Client(args.host, args.port)
        _client.register_message_handler(_message_handler)
        await _client.run()

        _token = args.token

        async def get_info_loop():
            while True:
                try:
                    await _get_info()
                except Exception as e:
                    _sdk_logger.error(f'Failed to get info: {e}')
                await asyncio.sleep(GET_INFO_INTERVAL)

        _task_list.append(asyncio.create_task(get_info_loop()))

    except Exception as e:
        _sdk_logger.error(f'Failed to initialize SDK: {e}')
        raise e


def get_agent() -> IAgent | None:
    """Gets the agent.

    Returns:
        The agent. None if the agent is not initialized.
    """
    return _agent


def get_blocks() -> IBlockSource | None:
    """Gets the block source.

    Returns:
        The block source. None if the block source is not initialized.
    """
    return _block_source


def get_entities() -> IEntitySource | None:
    """Gets the entity source.

    Returns:
        The entity source. None if the entity source is not initialized.
    """
    return _entity_source


def get_logger() -> ILogger:
    """Gets the user logger.

    Returns:
        The user logger.
    """
    return _user_logger


def get_tick() -> int | None:
    """Gets the current tick.

    Returns:
        The tick. None if the tick is not initialized.
    """
    if _tick_info is None:
        return None

    if _tps is None:
        return _tick_info.tick

    seconds_since_last_tick = (
        datetime.now() - _tick_info.time).total_seconds()
    estimated_ticks_since_last_tick = int(seconds_since_last_tick * _tps)

    return _tick_info.tick + estimated_ticks_since_last_tick


def get_tps() -> float | None:
    """Gets the current TPS.

    Returns:
        The TPS. None if the TPS is not initialized.
    """
    return _tps


def _message_handler(message: Message) -> None:
    try:
        global _block_source, _entity_source, _agent, _tick_info, _tps

        if message.get_bound_to() != Message.BoundToKind.CLIENT_BOUND.value:
            return

        if message.get_type() == Message.MessageKind.ERROR.value:
            _sdk_logger.error(
                f'The server returned an error: {message.get_obj()["message"]}')

        elif message.get_type() == Message.MessageKind.AFTER_BLOCK_CHANGE.value:
            if _block_source is None:
                return

            for block_change in message.get_obj()['change_list']:
                position = Position[int](block_change['position']['x'],
                                         block_change['position']['y'],
                                         block_change['position']['z'])
                _block_source[position] = Block(type_id=block_change['block_type_id'],
                                                 position=position)

        elif message.get_type() == Message.MessageKind.AFTER_ENTITY_CREATE.value:
            if _entity_source is None:
                return

            for creation in message.get_obj()['creation_list']:
                _entity_source.add_entity(Entity(type_id=creation['entity_type_id'],
                                                  unique_id=creation['unique_id'],
                                                  position=Position[float](creation['position']['x'],
                                                                           creation['position']['y'],
                                                                           creation['position']['z']),
                                                  orientation=Orientation[float](creation['orientation']['yaw'],
                                                                                 creation['orientation']['pitch'])))

        elif message.get_type() == Message.MessageKind.AFTER_ENTITY_ORIENTATION_CHANGE.value:
            if _entity_source is None:
                return

            for orientation_change in message.get_obj()['change_list']:
                entity = _entity_source[orientation_change['unique_id']]
                if entity is None:
                    continue

                entity.set_orientation(Orientation[float](orientation_change['orientation']['yaw'],
                                                          orientation_change['orientation']['pitch']))

                if _agent is not None and _agent.get_unique_id() == entity.get_unique_id():
                    _agent.set_orientation(entity.get_orientation())

        elif message.get_type() == Message.MessageKind.AFTER_ENTITY_POSITION_CHANGE.value:
            if _entity_source is None:
                return

            for position_change in message.get_obj()['change_list']:
                entity = _entity_source[position_change['unique_id']]
                if entity is None:
                    continue

                entity.set_position(Position[float](position_change['position']['x'],
                                                    position_change['position']['y'],
                                                    position_change['position']['z']))

                if _agent is not None and _agent.get_unique_id() == entity.get_unique_id():
                    _agent.set_position(entity.get_position())

        elif message.get_type() == Message.MessageKind.AFTER_ENTITY_REMOVE.value:
            if _entity_source is None:
                return

            for removal in message.get_obj()['removal_id_list']:
                _entity_source.remove_entity(removal)

        elif message.get_type() == Message.MessageKind.AFTER_PLAYER_INVENTORY_CHANGE.value:
            if _agent is None:
                return

            for change in message.get_obj()['change_list']:
                if change['player_unique_id'] is not _agent.get_unique_id():
                    continue

                for slot_change in change['change_list']:
                    _agent.get_inventory()[slot_change['slot']] = None if slot_change['item_type_id'] is None \
                        else ItemStack(type_id=slot_change['item_type_id'],
                                        count=slot_change['count'])

        elif message.get_type() == Message.MessageKind.GET_BLOCKS_AND_ENTITIES.value:
            _block_source = BlockSource()
            _entity_source = EntitySource()

            for section in message.get_obj()['sections']:
                _block_source.add_section(Section(position=Position[int](section['position']['x'],
                                                                          section['position']['y'],
                                                                          section['position']['z']),
                                                   block_id_list=section['blocks']))

            for entity in message.get_obj()['entities']:
                _entity_source.add_entity(Entity(type_id=entity['type_id'],
                                                  unique_id=entity['unique_id'],
                                                  position=Position[float](entity['position']['x'],
                                                                           entity['position']['y'],
                                                                           entity['position']['z']),
                                                  orientation=Orientation[float](entity['orientation']['yaw'],
                                                                                 entity['orientation']['pitch'])))

                if _agent is not None and _agent.get_unique_id() == entity['unique_id']:
                    _agent.set_position(Position[float](entity['position']['x'],
                                                        entity['position']['y'],
                                                        entity['position']['z']))
                    _agent.set_orientation(Orientation[float](entity['orientation']['yaw'],
                                                              entity['orientation']['pitch']))

        elif message.get_type() == Message.MessageKind.GET_PLAYER_INFO.value:
            if _client is None:
                return
            if _token is None:
                return

            player_info = message.get_obj()
            if _agent is None:
                # Create a new agent
                _agent = Agent(client=_client,
                                token=_token,
                                unique_id=player_info["unique_id"],
                                position=Position[float](
                                    player_info["position"]['x'],
                                    player_info["position"]['y'],
                                    player_info["position"]['z']
                                ),
                                orientation=Orientation[float](
                                    player_info['orientation']['yaw'],
                                    player_info['orientation']['pitch']
                                )
                                )

            # Update Info
            _agent.set_health(player_info['health'])
            _agent.set_position(Position[float](
                player_info["position"]['x'],
                player_info["position"]['y'],
                player_info["position"]['z']
            ))
            _agent.set_orientation(Orientation[float](player_info['orientation']['yaw'],
                                                      player_info['orientation']['pitch']))
            # Main slots
            _agent.get_inventory()._main_hand_slot = int(
                player_info['main_hand'])
            # inventory
            for slot in range(Inventory._SLOTS):
                item_stack = player_info['inventory'][slot]
                if item_stack is not None:
                    _agent.get_inventory()[slot] = ItemStack(
                        item_stack['type_id'], item_stack['count'])

        elif message.get_type() == Message.MessageKind.GET_TICK.value:
            tick_message = message.get_obj()
            _tick_info = _TickInfo(tick_message['tick'], datetime.now())
            _tps = tick_message['ticks_per_second']

    except Exception as e:
        _sdk_logger.error(f'Error while handling message: {e}')


async def _get_info():
    global _client
    if _client is None:
        return

    # Send get tick message
    _client.send(Message({
        "bound_to": Message.BoundToKind.SERVER_BOUND.value,
        "type": Message.MessageKind.GET_TICK.value,
        "token": _token,
    }))

    # Send get player info message
    _client.send(Message({
        "bound_to": Message.BoundToKind.SERVER_BOUND.value,
        "type": Message.MessageKind.GET_PLAYER_INFO.value,
        "token": _token,
    }))

    # Compute near section
    if _agent is not None:
        agent_position = _agent.get_position()
        agent_position = _agent.get_position()

        request_section_list = [{
            'x': int(agent_position.x + x*16),
            'y': int(agent_position.y + y*16),
            'z': int(agent_position.z + z*16)
        } for x in range(-1, 2) for y in range(-1, 2) for z in range(-1, 2)]

        # Send get blocks and entities message
        _client.send(Message({
            "bound_to": Message.BoundToKind.SERVER_BOUND.value,
            "type": Message.MessageKind.GET_BLOCKS_AND_ENTITIES.value,
            "token": _token,
            "request_section_list": request_section_list})
        )
